File: android/src/test/java/to/holepunch/bare/android/DefaultHostPluginsTest.kt
package to.holepunch.bare.android

import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class DefaultHostPluginsTest {
    private fun dispatchSync(
        registry: HostPluginRegistry,
        pluginId: String,
        event: String,
        args: JSONObject? = null,
    ): HostPluginRegistry.HostInvokeOutcome {
        val latch = CountDownLatch(1)
        var captured: HostPluginRegistry.HostInvokeOutcome? = null
        registry.dispatch(pluginId, event, args, null) { outcome ->
            captured = outcome
            latch.countDown()
        }
        assertTrue("handler did not respond", latch.await(5, TimeUnit.SECONDS))
        return captured ?: error("no outcome captured")
    }

    private fun statusOf(outcome: HostPluginRegistry.HostInvokeOutcome): String {
        assertTrue("should be Ok", outcome is HostPluginRegistry.HostInvokeOutcome.Ok)
        return (outcome as HostPluginRegistry.HostInvokeOutcome.Ok).value.getString("status")
    }

    @Test
    fun `unknown permission ids are never granted`() {
        val registry = HostPluginRegistry()
        registerDefaultHostPlugins(registry)

        for (event in listOf("permissions.request", "permissions.status")) {
            for (permission in listOf("location", "camera-x", "contacts", "")) {
                val outcome = dispatchSync(
                    registry,
                    "core.permissions",
                    event,
                    JSONObject().put("permission", permission),
                )
                assertEquals("$event $permission", "unsupported", statusOf(outcome))
            }
        }
    }

    @Test
    fun `storage needs no runtime grant on this host`() {
        val registry = HostPluginRegistry()
        registerDefaultHostPlugins(registry)

        for (event in listOf("permissions.request", "permissions.status")) {
            val outcome = dispatchSync(
                registry,
                "core.permissions",
                event,
                JSONObject().put("permission", "storage"),
            )
            assertEquals("$event storage", "granted", statusOf(outcome))
        }
    }

    @Test
    fun `camera without a context is not assumed granted`() {
        val registry = HostPluginRegistry()
        registerDefaultHostPlugins(registry)

        val outcome = dispatchSync(
            registry,
            "core.permissions",
            "permissions.status",
            JSONObject().put("permission", "camera"),
        )
        // No Context in a unit test; the real grant state is unknown, so this
        // must fail closed (never a fabricated grant).
        assertEquals("denied", statusOf(outcome))
    }

    @Test
    fun `capabilities list request and status`() {
        val registry = HostPluginRegistry()
        registerDefaultHostPlugins(registry)

        val json = registry.toCapabilitiesJson()
        assertEquals(1, json.length())
        val permissions = json.getJSONObject(0)
        assertEquals("core.permissions", permissions.getString("pluginId"))
        val events = permissions.getJSONArray("events")
        assertEquals("permissions.request", events.getString(0))
        assertEquals("permissions.status", events.getString(1))
    }
}

File: ios/Tests/BareHostTests/DefaultHostPluginsTest.swift
import XCTest
import BareHost

final class DefaultHostPluginsTest: XCTestCase {
  private func dispatchSync(
    _ registry: HostPluginRegistry,
    pluginId: String,
    event: String,
    args: [String: Any]? = nil
  ) -> HostPluginRegistry.HostInvokeOutcome {
    let expectation = expectation(description: "dispatch responded")
    var captured: HostPluginRegistry.HostInvokeOutcome?
    registry.dispatch(
      pluginId: pluginId,
      event: event,
      args: args,
      payload: nil
    ) { outcome in
      captured = outcome
      expectation.fulfill()
    }
    wait(for: [expectation], timeout: 5)
    return captured ?? .fail(code: "TEST", message: "no outcome")
  }

  func testStorageIsGrantedForStatusAndRequest() {
    let registry = HostPluginRegistry()
    registerDefaultHostPlugins(registry)

    for event in ["permissions.status", "permissions.request"] {
      let outcome = dispatchSync(
        registry,
        pluginId: "core.permissions",
        event: event,
        args: ["permission": "storage"]
      )
      guard case .ok(let value) = outcome else {
        return XCTFail("expected Ok outcome for \(event)")
      }
      XCTAssertEqual("storage", value["permission"] as? String)
      XCTAssertEqual("granted", value["status"] as? String)
    }
  }

  func testCameraStatusMapsToARealSystemState() {
    let registry = HostPluginRegistry()
    registerDefaultHostPlugins(registry)

    let outcome = dispatchSync(
      registry,
      pluginId: "core.permissions",
      event: "permissions.status",
      args: ["permission": "camera"]
    )
    guard case .ok(let value) = outcome else {
      return XCTFail("expected Ok outcome")
    }
    XCTAssertEqual("camera", value["permission"] as? String)
    let status = try? XCTUnwrap(value["status"] as? String)
    XCTAssertTrue(
      ["granted", "denied", "notDetermined", "unsupported"].contains(status ?? ""),
      "unexpected camera status \(status ?? "nil")"
    )
  }

  func testCapabilitiesListRequestAndStatus() {
    let registry = HostPluginRegistry()
    registerDefaultHostPlugins(registry)

    let rows = registry.toCapabilitiesJSON()
    XCTAssertEqual(1, rows.count)
    let permissions = try? XCTUnwrap(rows.first)
    XCTAssertEqual("core.permissions", permissions?["pluginId"] as? String)
    XCTAssertEqual(
      ["permissions.request", "permissions.status"],
      permissions?["events"] as? [String]
    )
  }
}
